### DISCUSSION TITLE
Configuration management system usage

### DISCUSSION THREAD

**Discussion by SchoolGuy** [cobbler/cobbler#3416] [Polls]
Currently, Cobbler offers basic built-in configuration management via Koan. This is functioning via the Package, File and Repo resources.

I am considering removing the code associated with this functionality for 4.0.0 (or 5.0.0) because I would like Cobbler to focus on being a good auto-installation server.

The reasoning behind this is that Cobbler already has integrations into other very mature configuration management systems like Ansible, Salt, Puppet and Chef. They are far more flexible and powerful than the built-in option. At the moment I have no intention - and manpower - to put focus on the built-in configuration management system.

If the interest of the community is big enough then of course I can change that, so please feel free to vote!

**Comment by opoplawski** [comment]
I guess I'm a bit unclear on the use-cases here, and I'm not sure what kind of "configuration management" is done with koan.  I'm a pretty heavy user of koan for deploying virtual machines and I use cobbler has a source of truth for my configuration management in ansible.  So I'd be very unhappy if `koan --virt` went away, and in fact would like to see it improved.

**Reply by SchoolGuy** [reply]
I am talking about the following bit of code here: https://github.com/cobbler/koan/blob/main/koan/configurator.py

I haven't gotten into the depths of what flags are responsible but essentially the virtualization should stay and will be improved once cloud-init is finally done.

**Comment by opoplawski** [comment]
Ah, never noticed that before.   Yeah, that should just go.

**Comment by SchoolGuy** [comment]
We have six votes that are for removing this functionality and no votes for the other options. As such I will close this poll with the result that in 3.4.0 this functionality will be removed. Since this is a breaking change and 3.4.0 will gain cloud-init support I will release 3.4.0 as 4.0.0.

**Comment by opoplawski** [comment]
So, apparently I did not fully understand what all was planning on being removed from cobbler.  I actually am an extremely heavy user of management classes.  None of the other components of a management class other than the name though.  We map this in the cobbler ansible plugin to groups for the the systems to belong to.

I don't mind changing the name but I think it would be very helpful to be able to assign systems in cobbler to groups.

**Reply by SchoolGuy** [reply]
Okay but this has nothing to do with the code that was behind that per-se. I also believe that system groups are already possible via sub-profiles. Or does the Code behind the management classes do anything for your ansible integration?

**Reply by opoplawski** [reply]
Management classes allowed you to assign multiple classes to a system.  This allows us to express the different roles that a machine might serve.  For example:

- dnsserver
- amanda_target
- printserver

A system can only have a single profile so it doesn't work for this.  Having the ability for cobbler to assign a system to multiple classes or groups is very useful.  It allows you to track that info where you define the systems.  Otherwise one needs to maintain a separate inventory file listing the groups a system belongs to and keep that in sync when systems come and go.  I don't need any of the other attributes a management class had, just the name.

**Reply by SchoolGuy** [reply]
I see the problem now. This is again not linked to the code that was present in Cobbler. This makes the class hierarchy a lot more complicated. I will see if we can at least partly revert that functionality. My main issue is that it is really hard for me to justify the maintenance of code that seems to serve no purpose and is not documented how it is supposed to work.

**Reply by opoplawski** [reply]
I just wanted to bring this up again as it is core to how we make use of cobbler.  What can I do to help make sure this (the aability to assign systems to multiple groups or "classes" (the name does not really matter)  does not go away?

**Reply by SchoolGuy** [reply]
After I have finished the Cloud Init PR, I can revisit this. As a stopgap measure you can always use autoinstall meta and make use of that inside the templates. Be assured, that I will not publish a release before your concerns are solved.

**Reply by SchoolGuy** [reply]
@opoplawski So the Cloud-Init and Agama PRs are now included in the development branch. I have written a blog post about the Chef integration, which will be merged soon. Would you be willing to set up a short call in the near future to discuss the next steps? I really wouldn't want to disrupt your use case.

**Reply by SchoolGuy** [reply]
I am not a fan of it, but technically, I can try to reintroduce the feature. The issue really is for me that I want to keep Cobbler as an installation server and would like to prevent the configuration management feature "Files" from being present in Cobbler. However, I found a use case from CERN that seems to make use of Cobbler in combination with Puppet. I have written to the author of the presentation to see if the setup still exists and how well they make use of it. If CERN makes use of this feature, then I will definitely need to re-add this if they ask me to do so.

**Reply by opoplawski** [reply]
@SchoolGuy definitely.  My needs at the moment are very limited - simply the ability to apply a management class to a system.  This essentially adds the system to an ansible "group".  I would be happy with cobbler simply having the ability to create "groups" and add systems to them.  None of the other stuff in "resources" is of use.

I can understand the desire to focus on cobbler as an install server.  But it is very useful as a source of "truth" about the network - operating system / version / network configuration / etc.  That would be a pain to have to duplicate in another system above it.

Another possible "wishlist" which I can see being seen as a bit more creep would be the ability to tag system with more arbitrary key / value pairs.  But I haven't really fleshed that out in my mind yet.

I'll try to get a cobbler dev system spun up soon so I can see what things look like currently.

**Reply by SchoolGuy** [reply]
@opoplawski I am open to a call if it helps you. My thought about creating groups would be subprofiles. Arbitrary key value pairs are IMHO covered via autoinstall meta. :)

**Reply by opoplawski** [reply]
subprofiles don't work because we need a one -> many relationship.  A system can be in an arbitrary number of groups.

**Reply by SchoolGuy** [reply]
Ah, that makes sense. So, creating a new Item called `MachineGroup` would be a possible solution that works for you?

**Reply by opoplawski** [reply]
I think that would be just what would be needed.  Thanks!

**Reply by SchoolGuy** [reply]
Let's discuss the required fields and scope here, please: #3983 I would like to start pushing for a feature freeze for 4.0.0, so I can start adjusting the clients and start testing the Cloud-Init and Agama integrations internally soon.
